src/k2/module/k23si/Persistence.cpp

```diff
@@ -29,54 +29,87 @@ Persistence::Persistence() {
     }
 
-    // In this method we have to execute a bunch of async steps:
-    // 1. persist the current batch (1 op)
-    // 2. notify the writers that their writes completed (N concurrent ops)
-    // As 1 and 2 above are happening, there could be new modifications arriving (e.g. new WIs being
-    // placed, or even writes issued by the continuations on the promises satisfied by 2).
-    // To prevent all of these cases of concurrent modification of the buffers and promises,
-    // we swap those out here, forcing any new modifications to be appended to a new batch.
-
-    // swap out the pending promises, clearing the active batch promises
-    std::vector<seastar::promise<>> proms;
-    proms.swap(_pendingProms);
-
     // move the buffered data into a single request and delete the buffer.
     // Any writes after this point will be appended to a new buffer/batch
     dto::K23SI_PersistenceRequest<Payload> request{};
     request.value.val = std::move(*_buffer);
     _buffer.reset(nullptr);
+    std::vector<seastar::promise<Status>> proms; // ditto for the pending promises
+    proms.swap(_pendingProms);
 
-    return RPC().callRPC<dto::K23SI_PersistenceRequest<Payload>, dto::K23SI_PersistenceResponse>
-        (dto::Verbs::K23SI_Persist, request, *_remoteEndpoint, _config.persistenceTimeout())
-        .then_wrapped([proms=std::move(proms)] (auto&& fut) mutable {
-        K2LOG_D(log::skvsvr, "flushed. Notifying {} pending promises", proms.size());
-
-            if (fut.failed()) {
-                auto exc = fut.get_exception();
-                for (auto&prom: proms) prom.set_exception(exc);
-                return seastar::make_exception_future<Status>(exc);
+    _flushFut = _flushFut
+       .then([this, request=std::move(request)] (auto&& status) mutable {
+            if (!status.is2xxOK()) {
+                // previous flush did not succeed. just pass this along
+                K2LOG_D(log::skvsvr, "previous flush was unsuccessful with status {}", status);
+                return RPCResponse(std::move(status), dto::K23SI_PersistenceResponse{});
             }
-
-            for (auto&prom: proms) prom.set_value();
-            return seastar::make_ready_future<Status>(std::get<0>(fut.get()));
+            K2LOG_D(log::skvsvr, "sending new flush");
+
+            return RPC().callRPC<dto::K23SI_PersistenceRequest<Payload>, dto::K23SI_PersistenceResponse>
+            (dto::Verbs::K23SI_Persist, request, *_remoteEndpoint, _config.persistenceTimeout());
+       })
+       .then([proms=std::move(proms)] (auto&& result) mutable {
+            auto& [status, _] = result;
+            // extract the RPC status and send it down the chain
+            K2LOG_D(log::skvsvr, "flushed with status: {}. Notifying {} promises", status, proms.size());
+            for (auto& prom: proms) {
+                prom.set_value(status);
+            }
+            return seastar::make_ready_future<Status>(std::move(status));
```

Comment:
I think there should be a handle_exception block here that fulfills the promises with exceptions.

Reply:
Added then_wrapped to handle exceptions